andersoncd/weighted_lasso.py

```diff
@@ -0,0 +1,394 @@
+    """
+
+    if sparse.issparse(X):
+        raise ValueError("Spare design matrices are not supported yet.")
+
+    X = check_array(X, 'csc', dtype=[np.float64, np.float32],
+                    order='F', copy=False, accept_large_sparse=False)
+    y = check_array(y, 'csc', dtype=X.dtype.type, order='F', copy=False,
+                    ensure_2d=False)
+
+    n_samples, n_features = X.shape
+
+    # if X_offset is not None:
+    #     X_sparse_scaling = X_offset / X_scale
+    #     X_sparse_scaling = np.asarray(X_sparse_scaling, dtype=X.dtype)
+    # else:
+    #     X_sparse_scaling = np.zeros(n_features, dtype=X.dtype)
+
+    # X_dense, X_data, X_indices, X_indptr = _sparse_and_dense(X)
+
+    if weights is None:
+        weights = np.ones(n_features, dtype=X.dtype)
+    elif (weights < 0).any():
+        raise ValueError("Strictly negative weights are not supported.")
+
+    penalized = weights > 0
+    if alphas is None:
+        alpha_max = np.max(np.abs(
+            X[:, penalized] @ y / weights[penalized])) / n_samples
+
+        alphas = alpha_max * np.geomspace(1, eps, n_alphas, dtype=X.dtype)
+    else:
+        alphas = np.sort(alphas)[::-1]
+
+    n_alphas = len(alphas)
+
+    coefs = np.zeros((n_features, n_alphas), order='F', dtype=X.dtype)
+    kkt_maxs = np.zeros(n_alphas)
+
+    if return_n_iter:
+        n_iters = np.zeros(n_alphas, dtype=int)
+
+    norms_X_col = norm(X, axis=0)
+
+    for t in range(n_alphas):
+        alpha = alphas[t]
+        if verbose:
+            to_print = "##### Computing alpha %d/%d" % (t + 1, n_alphas)
+            print("#" * len(to_print))
+            print(to_print)
+            print("#" * len(to_print))
+        if t > 0:
+            w = coefs[:, t - 1].copy()
+            p0 = max(len(np.where(w != 0)[0]), 1)
+        else:
+            if coef_init is not None:
+                w = coef_init.copy()
+                p0 = max((w != 0.).sum(), p0)
+                R = y - X @ w
+            else:
+                w = np.zeros(n_features, dtype=X.dtype)
+                R = y.copy()
+
+        sol = celer_primal(
+            X, y, alpha, w, R, norms_X_col, weights,
+            max_iter=max_iter, max_epochs=max_epochs, p0=p0, tol=tol,
+            verbose=verbose)
+
+        coefs[:, t] = w.copy()
+        kkt_maxs[t] = sol[-1]
+        # TODO sol[0], sol[1], sol[2][-1]
+        if return_n_iter:
+            n_iters[t] = len(sol[1])
+
+    results = alphas, coefs, kkt_maxs
+    if return_n_iter:
+        results += (n_iters,)
+
+    return results
+
+
+# def _kkt_violation(XtR, weights, alpha):
+#     for j in range(n_features):
+#         if weights[j] > 0:
+#             kkt[j] = max(0, np.abs(XtR[j]) / n_samples - alpha * weights[j]))
+#         else:
+#                 kkt_max=max(kkt_max, np.abs(XtR[j]))
+
+
+def celer_primal(
+        X, y, alpha, w, R, norms_X_col, weights, max_iter=50,
+        max_epochs=50_000, p0=10, tol=1e-4, use_acc=False, K=5, verbose=0):
```

Comment:
In the end in celer I unexposed it. K=5 can be hardcoded since it's a really robust choice